### PR TITLE
Changes in HGCal geometry vs flatfile comparison

### PR DESCRIPTION
**Pull request by yulunmiao**
#### PR description:

Changes are made in Validation/HGCalValidation/plugins/HGCalWaferValidation.cc  
to allow reading wafer information from new flat file format and compare to geometry. 
and in Validation/HGCalValidation/test/python/testHGCalWaferValidation_cfg.py to allow take geometry D86 as argument and can specify certain flat file via the path to it.

More details can be found here https://indico.cern.ch/event/1095051/contributions/4610617/attachments/2344534/3997691/GeometryD86vsflatfile_yulunmiao.pdf
#### PR validation:

Checks using geometry D83 with old flat file format and geometry D86 with new flat file format is done, no bug is found in code

### CODE REVIEW

**Comment by qliphy** [comment]
@yulunmiao  Please write a meaningful PR title.    Moreover, make a forward PR in master (12_2_X)

**Comment by yulunmiao** [comment]
Hi @qliphy , thank you for your comments, I have made the changes.

**Comment by perrotta** [comment]
> Hi @qliphy , thank you for your comments, I have made the changes.

@yulunmiao Since this PR moved to master, before we can test it you have now to apply the code-format fixes, as explained in https://github.com/cms-sw/cmssw/pull/36174#issuecomment-974097177

File: Validation/HGCalValidation/plugins/HGCalWaferValidation.cc [hunk 1]
```diff
@@ -206,7 +211,30 @@ void HGCalWaferValidation::ProcessWaferLayer(DDCompactView::GraphWalker& walker)
...
       //edm::LogVerbatim(logcat) << "rotStr " << rotStr << " rotCode " << rotCode;
-      waferInfo.shapeCode = shapeStr;
+
+      // convert shape code to wafer types defined in HGCalTypes.h
+      waferInfo.shapeCode = 99;
+      if (shapeStr == "F")
+        waferInfo.shapeCode = 0;
+      if (shapeStr == "a")
+        waferInfo.shapeCode = 4;
+      if (shapeStr == "am")
```
Comment (jfernan2) [inline Validation/HGCalValidation/plugins/HGCalWaferValidation.cc:221]: what prevents all these ifs to be else if?

File: Validation/HGCalValidation/plugins/HGCalWaferValidation.cc [hunk 2]
```diff
@@ -323,15 +351,90 @@ void HGCalWaferValidation::analyze(const edm::Event& iEvent, const edm::EventSet
...
+      if (waferDensityStr == "l") {
+        if (waferShapeStr == "0")
+          waferShapeCode = 0;
+        if (waferShapeStr == "1")
+          waferShapeCode = 4;
+        if (waferShapeStr == "2")
+          waferShapeCode = 4;
+        if (waferShapeStr == "3")
+          waferShapeCode = 5;
+        if (waferShapeStr == "4")
```
Comment (jfernan2) [inline Validation/HGCalValidation/plugins/HGCalWaferValidation.cc:386]: Same comment here

File: Validation/HGCalValidation/plugins/HGCalWaferValidation.cc [hunk 3]
```diff
@@ -323,15 +351,90 @@ void HGCalWaferValidation::analyze(const edm::Event& iEvent, const edm::EventSet
...
+      waferU = std::stoi(tokens[6]);
+      waferV = std::stoi(tokens[7]);
+      waferShapeCode = 99;
+      if (waferDensityStr == "l") {
+        if (waferShapeStr == "0")
+          waferShapeCode = 0;
+        if (waferShapeStr == "1")
+          waferShapeCode = 4;
+        if (waferShapeStr == "2")
+          waferShapeCode = 4;
```
Comment (jfernan2) [inline Validation/HGCalValidation/plugins/HGCalWaferValidation.cc:383]: Same assignment than 2 lines above?

File: Validation/HGCalValidation/plugins/HGCalWaferValidation.cc [hunk 4]
```diff
@@ -323,15 +351,90 @@ void HGCalWaferValidation::analyze(const edm::Event& iEvent, const edm::EventSet
...
+        if (waferShapeStr == "3")
+          waferShapeCode = 5;
+        if (waferShapeStr == "4")
+          waferShapeCode = 5;
```
Comment (jfernan2) [inline Validation/HGCalValidation/plugins/HGCalWaferValidation.cc:387]: Same assignment than two lines above, these ifs could be merged, right?

File: Validation/HGCalValidation/plugins/HGCalWaferValidation.cc [hunk 5]
```diff
@@ -323,15 +351,90 @@ void HGCalWaferValidation::analyze(const edm::Event& iEvent, const edm::EventSet
...
+        if (waferShapeStr == "0")
+          waferShapeCode = 0;
+        if (waferShapeStr == "1")
+          waferShapeCode = 8;
+        if (waferShapeStr == "2")
+          waferShapeCode = 3;
+        if (waferShapeStr == "3")
+          waferShapeCode = 6;
+        if (waferShapeStr == "4")
+          waferShapeCode = 6;
```
Comment (jfernan2) [inline Validation/HGCalValidation/plugins/HGCalWaferValidation.cc:402]: same comment, assignment is the same

File: Validation/HGCalValidation/plugins/HGCalWaferValidation.cc [hunk 6]
```diff
@@ -323,15 +351,90 @@ void HGCalWaferValidation::analyze(const edm::Event& iEvent, const edm::EventSet
...
+        waferShapeCode = 0;
+      if (waferShapeStr == "a")
+        waferShapeCode = 4;
+      if (waferShapeStr == "am")
+        waferShapeCode = 8;
+      if (waferShapeStr == "b")
+        waferShapeCode = 1;
+      if (waferShapeStr == "bm")
+        waferShapeCode = 9;
+      if (waferShapeStr == "c")
```
Comment (jfernan2) [inline Validation/HGCalValidation/plugins/HGCalWaferValidation.cc:427]: again, these ifs could be turned into else if rights?

**Comment by yulunmiao** [comment]
Hi @jfernan2 , thank you for your comments, I have changes the questions you mentioned

**Comment by jfernan2** [comment]
please test

File: Validation/HGCalValidation/plugins/HGCalWaferValidation.cc (hunk 1 above)
Comment (jfernan2) [reply Validation/HGCalValidation/plugins/HGCalWaferValidation.cc:221]: THese should be replaced by else if too

File: Validation/HGCalValidation/plugins/HGCalWaferValidation.cc (hunk 3 above)
Comment (jfernan2) [reply Validation/HGCalValidation/plugins/HGCalWaferValidation.cc:383]: Again

File: Validation/HGCalValidation/plugins/HGCalWaferValidation.cc (hunk 2 above)
Comment (jfernan2) [reply Validation/HGCalValidation/plugins/HGCalWaferValidation.cc:386]: Not migrated yet

File: Validation/HGCalValidation/plugins/HGCalWaferValidation.cc (hunk 4 above)
Comment (jfernan2) [reply Validation/HGCalValidation/plugins/HGCalWaferValidation.cc:387]: Same

File: Validation/HGCalValidation/plugins/HGCalWaferValidation.cc (hunk 6 above)
Comment (jfernan2) [reply Validation/HGCalValidation/plugins/HGCalWaferValidation.cc:427]: Revisit

**Comment by yulunmiao** [comment]
Hi @jfernan2, I am really sorry that I submitted the file into the wrong folder, I have changed it to the correct one now

**Comment by jfernan2** [comment]
please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by pfs** [comment]
@yulunmiao @bsunanda @imranyusuff  I have a request and a question

Request: in the mapping of strings to wafer types, can you use the names in the [enum](https://github.com/cms-sw/cmssw/blob/master/Geometry/HGCalCommonData/interface/HGCalTypes.h#L57) .
I.e. instead of lines like
```
if (shapeStr == "a")
        waferInfo.shapeCode = 4;
```
do something like
```
if (shapeStr == "a")
        waferInfo.shapeCode = WaferPartialType::WaferHalf;
```
etc. which is easier for lay people to understand and not get confused by the fact that a different numbering scheme is used outside CMSSW (cf [slide 3](https://indico.cern.ch/event/1085972/contributions/4567337/attachments/2329012/3968212/GeomIssues.pdf)).

**Comment by yulunmiao** [comment]
@pfs Thanks for you comment. I've made the changes

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**